File: csharp/AkkaBank/BankAccountActor.cs
using System;
using Akka.Actor;

namespace AkkaBank
{
    public class BankAccountActor : ReceiveActor
    {
        private double balance;

        public BankAccountActor()
        {
            Become(Opened);
        }

        private void Opened()
        {
            Receive<GetBalance>(msg => Sender.Tell(balance));

            Receive<Deposit>(msg => balance += msg.Amount);

            Receive<Withdrawal>(msg =>
            {
                if (balance < msg.Amount)
                {
                    throw new Exception("Insufficient funds");
                }
                balance -= msg.Amount;
            });

            Receive<Close>(msg => Become(Closed));
        }

        private void Closed()
        {
            Receive<GetBalance>(msg => Sender.Tell(balance));

            Receive<Open>(msg => Become(Opened));

            ReceiveAny(msg =>
            {
                throw new Exception("Account is closed");
            });
        }
    }
}

File: scala/src/main/scala/BankAccountActor.scala
import akka.actor.Actor

class BankAccountActor extends Actor {
  var balance : Double = 0

  def receive = opened

  val opened: Receive = {
    case GetBalance(_) => sender ! balance
    case Deposit(_, amount) => balance += amount
    case Withdrawal(_, amount) => {
      if (balance < amount) throw new Exception("Insufficient funds")
      balance -= amount
    }
    case Close => context.become(closed)
  }

  val closed: Receive = {
    case GetBalance => sender ! balance
    case Open => context.become(opened)
    case _ => throw new Exception("Account is closed")
  }
}
